### ISSUE TITLE
_FillValue not Autofilling Data

### ISSUE THREAD

**Issue by Elpolloloc0** [Unidata/netcdf4-python#774]
Hi,

When I try to create  a netCDF file and test _FillValue to automatically fill the data with the value set, nothing happens/I get incorrect values put in its place.
The first test is setting the _FillValue in the variable itself while the second I test it with setting it as an attribute myself, but both do not get the results I seek.
If I am doing something wrong in how, or when, I am supposed activate the _FillValue then I would greatly appreciate any help in this matter.

Script::
import netCDF4 as netCDF

fileout = netCDF.Dataset("testfill.nc", "w", format="NETCDF3_CLASSIC")

fileout.set_fill_on()
fileout.createDimension("TEST", 100)
fileout.createVariable("TESTNUM", "f8", "TEST", fill_value=-99.0)

fileout.createVariable("TESTNUM2", "f8", "TEST")
fileout.variables["TESTNUM2"].setncattr("_FillValue", -99.0)

fileout.close()

-------------------------------------------------------------------------------------------------------------------
Output::
netcdf testfill {
dimensions:
	TEST = 100 ;
variables:
	double TESTNUM(TEST) ;
		TESTNUM:_FillValue = -99. ;
	double TESTNUM2(TEST) ;
		TESTNUM2:_FillValue = -99. ;
data:

 TESTNUM = _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, 
    _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, 
    _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, 
    _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, 
    _, _, _, _, _, _ ;

 TESTNUM2 = 9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36, 9.96920996838687e+36, 9.96920996838687e+36, 
    9.96920996838687e+36 ;
}
 
I am currently using the up to date netCDF4 module in Python 3.4.

**Comment by jswhit** [comment]
The fill value must be set a variable creation time.  In your first example, when you set it in createVariable, the variable is prefilled with -99, and when you print the variable those values show up as masked ('_').  In the second case, the variable was filled with the default fill value (9.96920996838687e+36) since you did not specify it in createVariable.  Setting at after the effect doesn't do anything since the variable was already pre-filled with the default value.

Bottom line - specifying it as a kwarg in createVariable is the right way to do it.  I'm not sure what about the result you saw is not expected.

**Comment by Elpolloloc0** [comment]
Ok thank you for the reply!
I changed my code to see if I could unmask the values to see if it could work as shown below:

import netCDF4 as netCDF
filout = netCDF.Dataset("testfill.nc", "w", format="NETCDF3_CLASSIC")
filout.set_fill_on()
filout.createDimension("TEST", 100)
filout.createVariable("TESTNUM", "f8", "TEST", fill_value=-99.0)
filout.variables["TESTNUM"].set_auto_mask(False)
filout.close()

However this was my output:

netcdf testfill {
dimensions:
	TEST = 100 ;
variables:
	double TESTNUM(TEST) ;
		TESTNUM:_FillValue = -99. ;
data:

 TESTNUM = _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, 
    _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, 
    _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, 
    _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, 
    _, _, _, _, _, _ ;
}

For some reason the values are still masked.

**Comment by jswhit** [comment]
Are you showing the output of ncdump? If so, ncdump always masks the output values that are equal to the fill value. If you read the data from the variable in python and print the data, you will see the -99 values if you have set_auto_mask(False).

**Comment by Elpolloloc0** [comment]
That was the output in ncdump. I see the results now in Python. Thank you very much for the explanation and clarification on how it works,
